### ISSUE TITLE
Docker for mac commands extremely slow, even for basic commands

### ISSUE THREAD

**Issue by petems** [docker/for-mac#1385]
### Expected behavior

Normal response times for docker commands, both commands sent to containers and command line arguments.

### Actual behavior

Everything extremely slow. 

```
$ time docker --version
Docker version 17.03.0-ce, build 60ccb22

real	0m32.066s
user	0m0.042s
sys	0m0.019s
```

Commands that are lightning quick, when I use docker-machine with a virtualbox instance:

```
$ eval $(docker-machine env default)
$ time docker --version
Docker version 17.03.0-ce, build 60ccb22

real	0m0.087s
user	0m0.062s
sys	0m0.021s
```

### Information

  - Full output of the diagnostics from "Diagnose & Feedback" in the menu
```
Docker for Mac: version: 17.03.0-ce-mac1 (e5a07a1a6)
macOS: version 10.11.6 (build: 15G1217)
logs: /tmp/D79091A1-C5CC-4557-B68E-AB5D1FE037DF/20170304-151114.tar.gz
[OK]     vmnetd
[OK]     dns
[OK]     driver.amd64-linux
[OK]     virtualization VT-X
[OK]     app
[OK]     moby
[OK]     system
[OK]     moby-syslog
[OK]     db
[OK]     env
[OK]     virtualization kern.hv_support
[OK]     slirp
[OK]     osxfs
[OK]     moby-console
[OK]     logs
[OK]     docker-cli
[OK]     menubar
[OK]     disk
```

This appears to have started since the update to 17.03.0-ce, but I found my old `1.12.6.14937` release DMG, installed that and the same thing happened.

### Steps to reproduce the behavior

  1. Run any docker command such as `docker --version`
  2. Response takes a long time

**Comment by robertjpayne** [comment]
I'm having similar issues, doesn't appear that downgrading Docker helps at all. I feels like something got bugged up in a cache and isn't getting cleaned up even on an uninstall to fix itself.

**Comment by clok** [comment]
I upgraded to `Version 17.03.0-ce-mac2 (15654)` this morning and I am now seeing extremely slow run time, build time and overall performance has degraded.

**Comment by MagnusS** [comment]
Thanks for the report @petems. Could there be any left over `DOCKER_` variables from Docker Machine that could be causing this? To check this you could try to connect directly to the Docker for Mac socket with `docker -H unix:///var/run/docker.sock --version` to see if you get the same delays.

If that doesn't help, please try to upgrade to the latest version (17.03.0-ce-mac2) and upload new logs using Diagnose & Upload. 

(I removed the .zip file with the logs from this issue - we just need the ID to retrieve them)

@clock @robertjpayne It may help to connect directly to the socket - if not, please use "Diagnose and Feedback" from the menu to upload your logs and submit a new ticket so we can take a look at it! Thanks

**Comment by aj0408** [comment]
Updated to Docker version 17.03.0-ce, build 60ccb22. Having an issue creating docker image 
"java.lang.RuntimeException: Could not parse Version from 17.03.0-ce
: `.' expected but `3' found
"

Using sbt docker to build the image. Worked fine before the upgrade

**Comment by neksa** [comment]
I had similar issue. The reason was a previous installation of docker. Check your env with
set|grep DOCKER

**Comment by petems** [comment]
Ok, so I went full nuclear: 
* killed my brew install
* killed docker-mac
* docker-machine.

Did a completely new brew install, then installed docker-for-mac.

On a vanilla docker-for-mac install, `docker --version` was back to speedy times:

``` 
$ time docker --version
Docker version 17.03.0-ce, build 60ccb22

real	0m0.023s
user	0m0.010s
sys	0m0.011s
```

However, the commands on containers were still taking a long time. But it might've been an upstream issue with CentOS filesystems in the container: When I was testing I tried the ubuntu:16.04 image, it was speedy as normal. Using the official CentOS 7 docker image `centos:7`, and echo "Hello world" would take 4-10 seconds.

I switched my centos image from `centos:7` to `jdeathe/centos-ssh:centos-7` and it worked as expected, with commands taking a fraction of a second.

Doing a bit of googling, there's a few references to CentOS slowness with Docker, but this seems to refer to a CentOS host, rather than guest. 

So... bit of a hsienbug. It seems to work at the right speed using the new image.

**Comment by djs55** [comment]
I'm glad the issue with `docker --version` has gone away. Perhaps there was some strange mixup between a command from `brew` and from Docker for Mac? Mysterious.

While 32 seconds for `--version` is inexplicably slow, `4-10` seconds for a `docker run` seems like a different kind of problem. Unfortunately the bug report upload is incomplete so I can't tell what type of hardware you're using but there has been a problem in the past with spinning disks. This is a bit speculative but when `docker` (or some other process) tries to persist information it sends a `flush` command on the block device. We implemented this in the most safe way possible with the macOS `fcntl(F_FULLFSYNC)` API. However we've since discovered this can be extremely slow -- multiple flushes can add up. Perhaps try switching to faster `fsync` mode by:

```
$ cd ~/Library/Containers/com.docker.docker/Data/database/
$ git checkout master
$ git reset --hard
HEAD is now at f7f4b43 last-start-time changed at 1490350978
$ mkdir -p com.docker.driver.amd64-linux/disk/
$ echo os > com.docker.driver.amd64-linux/disk/on-flush
$ git add com.docker.driver.amd64-linux/disk/on-flush 
Davids-MBP-2:database djs$ git commit -s -m 'switch to fsync'
[master eb2dd22] switch to fsync
 1 file changed, 1 insertion(+)
 create mode 100644 com.docker.driver.amd64-linux/disk/on-flush
```
The VM will reboot at this point.

Let me know if this makes any difference at all.

**Comment by petems** [comment]
@djs55 Thanks for the quick response! 

For context, I'm using docker containers for testing in a very crude way: 
* create container from image
* install ssh and start it
* add an ssh key
* use `CMD ["/sbin/init"]` as the start command
* Treat container like a "normal" VM, SSHing into it rather than using the Docker API.

I know this is not the "proper" way of using docker, but it works for us 😅 

So I did your steps and added the `fsync` setting, but no luck, a simple command to check if curl is installed takes 20 seconds:

```
centos-7-x64 11:57:13$ rpm -q curl
  curl-7.29.0-35.el7.centos.x86_64

centos-7-x64 executed in 20.10 seconds
```

Switching to the community CentOS 7 image is super quick:

```
centos-7-x64 12:02:53$ rpm -q curl
  curl-7.29.0-35.el7.centos.x86_64

centos-7-x64 executed in 0.03 seconds
```

Is this something that's a known issue that I could take up in the CentOS upstream image? This only seems to affect docker-mac, it works as expected on my Ubuntu 16.04 machine at home...

**Comment by djs55** [comment]
@petems thanks for the update!

I've not seen this issue before but it does sound like the issue is somewhere in the official CentOS 7 image. To dig a little deeper you could try running something like `strace -ff rpm -q curl` in both environments and comparing the system calls emitted by both. Note the container would need to be started with `--privileged` for this to work. Perhaps also try comparing other commands: `rpm` is perhaps quite a complicated one. Maybe simple things like
```
sh-4.2# dd if=/dev/zero of=1GiB bs=1M count=1024
1024+0 records in
1024+0 records out
1073741824 bytes (1.1 GB) copied, 2.85822 s, 376 MB/s
sh-4.2# time sha1sum 1GiB 
2a492f15396a6768bcbca016993f4b4c8b0b5307  1GiB

real	0m1.477s
user	0m1.390s
sys	0m0.080s
```
It's also worth checking in `top` that there isn't some other activity using all the CPU/ disk bandwidth. Do you have any volume shares?

I recommend doing a few experiments with `strace` and then reporting this upstream to the maintainer of the official CentOS image. Perhaps some optimisation was added to the community image somehow?

To summarise then
- the original `docker --version` slowness remains a mystery but at least has gone away by itself
- the `rpm -q` slowness seems to be problem specific to the image

Since I think the problem is to do with the image I'll close this issue for now. Please do let me know if new evidence comes to light which suggests a docker problem.

(If anyone else has a similar-sounding problem, could you create a fresh ticket with fresh diagnostics?)

Thanks for your report!

**Comment by TecnoSigma** [comment]
I had the same problem.

I solved with the fix below. On the Mac:

`System Preferences >> Network >> Advanced >> DNS Option`

If so, you have more than one DNS server, just leave the one that keeps your Docker faster.

With me it worked.


Good luck!

**Comment by Tkwon123** [comment]
Same problem here.  Unsetting all of my previous docker ENV tied to my docker-machine did the trick

**Comment by dumitrescustefan** [comment]
I had the same problem, solved it by manually setting a single DNS, like @TecnoSigma suggested. I figured it was the DNS looking at the timeouts in the "Diagnose & Feedback" menu. Did a bit of searching to find this thread, but now everything's snappy again. Tx!

**Comment by ealbert** [comment]
In my case, after doing an upgrade from an old version, the docker commands in the terminal were very slow. I checked my .bash_profile file and I found four env variables related to Docker; I removed them and that resolved my issue.
